sky/skylet/log_lib.py

```diff
@@ -440,15 +476,39 @@ def tail_logs(job_id: Optional[int],
         with open(log_path, 'r', newline='', encoding='utf-8') as log_file:
             # Using `_follow` instead of `tail -f` to streaming the whole
             # log and creating a new process for tail.
+            if tail > 0:
+                current_file_lines = log_file.readlines()
+                lines = collections.deque(current_file_lines, maxlen=tail)
+                start_stream = _should_start_streaming(current_file_lines,
+                                                       lines, start_stream_at)
+                for line in lines:
+                    if start_stream_at in line:
+                        start_stream = True
+                    if start_stream:
+                        print(line, end='')
+                # Flush the last n lines
+                print(end='', flush=True)
+            # Now, the cursor is at the end of the last lines
+            # if tail > 0
             for line in _follow_job_logs(log_file,
                                          job_id=job_id,
+                                         start_stream=start_stream,
                                          start_streaming_at=start_stream_at):
                 print(line, end='', flush=True)
     else:
         try:
             start_stream = False
             with open(log_path, 'r', encoding='utf-8') as f:
-                for line in f.readlines():
+                if tail > 0:
+                    # If tail > 0, we need to read the last n lines.
+                    # We use double ended queue to rotate the last n lines.
+                    current_file_lines = f.readlines()
+                    lines = collections.deque(current_file_lines, maxlen=tail)
+                    start_stream = _should_start_streaming(
+                        current_file_lines, lines, start_stream_at)
```

Comment:
With the function above this will become
```suggestion
                    # If tail > 0, we need to read the last n lines.
                    # We use double ended queue to rotate the last n lines.
                    _print_lines_after_start_stream_at(log_file.readlines(), start_stream_at, tail)
                    return
```

Reply:
Same as above reason, don't have the whole file in memory now